docs/protocols/xlang_object_graph_spec.md

```diff
@@ -0,0 +1,671 @@
+
+#### signed int32
+
+- size: 1~5 byte
+- Format: First convert the number into positive unsigned int by `(v << 1) ^ (v >> 31)` ZigZag algorithm, then encode
+  it as an unsigned varint.
+
+#### unsigned int64
+
+- size: 1~9 byte
+- Fury SLI(Small long as int) Encoding:
+    - If long is in `[0, 2147483647]`, encode as 4 bytes int: `| little-endian: ((int) value) << 1 |`
+    - Otherwise write as 9 bytes: `| 0b1 | little-endian 8 bytes long |`
+- Fury PVL(Progressive Variable-length Long) Encoding:
+    - positive long format: first bit in every byte indicates whether to have the next byte. If first bit is set
+      i.e. `b & 0x80 == 0x80`, then the next byte should be read until the first bit is unset.
+
+#### signed int64
+
+- size: 1~9 byte
+- Fury SLI(Small long as int) Encoding:
+    - If long is in `[-1073741824, 1073741823]`, encode as 4 bytes int: `| little-endian: ((int) value) << 1 |`
+    - Otherwise write as 9 bytes: `| 0b1 | little-endian 8 bytes long |`
+- Fury PVL(Progressive Variable-length Long) Encoding:
+    - First convert the number into positive unsigned long by `(v << 1) ^ (v >> 63)` ZigZag algorithm to reduce cost of
+      small negative numbers, then encoding it as an unsigned long.
+
+#### float32
+
+- size: 4 byte
+- format: encode the specified floating-point value according to the IEEE 754 floating-point "single format" bit layout,
+  preserving Not-a-Number (NaN) values, then write as binary by little endian order.
+
+#### float64
+
+- size: 8 byte
+- format: encode the specified floating-point value according to the IEEE 754 floating-point "double format" bit layout,
+  preserving Not-a-Number (NaN) values. then write as binary by little endian order.
+
+### string
+
+Format:
+
+```
+| header: size << 2 | 2 bits encoding flags | binary data |
+```
+
+- `size + encoding` will be concat as a long and encoded as an unsigned var long. The little 2 bits is used for
+  encoding:
+  0 for `latin`, 1 for `utf-16`, 2 for `utf-8`.
+- encoded string binary data based on encoding: `latin/utf-16/utf-8`.
+
+Which encoding to choose:
+
+- For JDK8: fury detect `latin` at runtime, if string is `latin` string, then use `latin` encoding, otherwise
+  use `utf-16`.
+- For JDK9+: fury use `coder` in `String` object for encoding, `latin`/`utf-16` will be used for encoding.
+- If the string is encoded by `utf-8`, then fury will use `utf-8` to decode the data. But currently fury doesn't enable
+  utf-8 encoding by default for java. Cross-language string serialization of fury uses `utf-8` by default.
+
+### list
+
+Format:
+
+```
+length(unsigned varint) | elements header | elements data
+```
+
+#### elements header
+
+In most cases, all elements are same type and not null, elements header will encode those homogeneous
+information to avoid the cost of writing it for every element. Specifically, there are four kinds of information
+which will be encoded by elements header, each use one bit:
```

Comment:
Can `length+elements` also be concat as a long and encoded as an unsigned var long like string type?

Reply:
Good suggestion! It's a good optimization for small collection.

But elements header took 4 bits, only 3 bits in a varint of a byte can be used for encoding length, the max value for it is `7`. Then we will use two bytes for length encoding, and the byte saving with this trick will be ineffective for lsit with size larger than `7`

I don't know the common size for collection, so not sure whether we should add such optimization. Either those bits left empty for future extension, or for this optimization.

@PragmaTwice @theweipeng Do you have any suggestions?

Reply:
I think we can concate length+elements header together

Reply:
@chaokunyang  If it is larger than 7, will the 3 bits be empty?

Reply:
Nope, it will be used, but next byte will have 3 bits empty

Reply:
I added this optimization to the spec since empty list and small list is common. thanks @LiangliangSui for this suggestion.